R/type-integer64.R

```diff
@@ -0,0 +1,142 @@
+
+#' 64 bit integers
+#'
+#' A `integer64` vector is a 64 bits integer vector. Details
+#' are implemented in the `bit64` package.
+#'
+#' These functions help the `integer64` class from `bit64` in to
+#' the vctrs type system by providing constructors, coercion functions,
+#' and casting functions. `new_int64()` is a low-level
+#' constructor that only checks the type is valid, so
+#' is for expert use only.
+#'
+#' @param x 64 bit integer vector
+#' @param ...,class Used to for subclasses.
+#' @keywords internal
+#' @export
+new_int64 <- function(x = bit64::integer64(), ..., class = character()) {
+  stopifnot(inherits(x, "integer64"))
+  x
+}
+
+# Print -------------------------------------------------------------------
+
+#' @rdname new_int64
+#' @export
+vec_ptype_full.integer64 <- function(x) {
+  "integer64"
+}
+
+#' @rdname new_int64
+#' @export
+vec_ptype_abbr.integer64 <- function(x) {
+  "int64"
+}
+
+
+# Coerce ------------------------------------------------------------------
+
+#' @export
```

Comment:
Can you please check this is up to date with the advice in https://vctrs.r-lib.org/articles/s3-vector.html#double-dispatch?

Reply:
I'm just unsure about 

```r
#' @export
#' @method vec_cast.integer64 logical
vec_cast.integer64.logical <- function(x, to) {
  bit64::as.integer64(x)
}
```

Does this need to distinguish the unspecified case given that cast to logical is allowed, or is this good enough: 

``` r
library(vctrs)
x <- bit64::as.integer64(1:10)
vec_cast(x, unspecified())
#> Error: Can't cast <integer64> to <vctrs_unspecified>
vec_cast(x, NA)
#>  [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
```

<sup>Created on 2018-11-20 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

Reply:
It's the other way around:

```R
library(vctrs)
x <- bit64::as.integer64(1:10)
vec_cast(unspecified(), x)
vec_cast(NA, x)
```

And I think you need to use `vec_unspecified_cast(x, to)` in that method